cirq/circuits/expectation_value.py

```diff
@@ -0,0 +1,376 @@
+                  paulistring2 = cirq.PauliString(qubit_pauli_map=
+                    {qubits[0]:cirq.Pauli.Z, qubits[2]: (cirq.Pauli.Y)})
+                  operator = {paulistring: 1.234, paulistring2: -5.678}
+        n_samples: number of runs for sampling
+
+    Returns:
+        expectation value of operator argument
+    """
+
+    qubits = list(circuit.all_qubits())
+    sim = XmonSimulator()
+
+    # Check whether circuit already has measurement gates:
+    last_moment_index = len(circuit) - 1
+    last_operations = circuit[last_moment_index].operations
+
+    measurement_qubits = list(qubits)
+    indices_measurement = list(range(len(qubits)))
+
+    for operation in last_operations:
+        if MeasurementGate.is_measurement(operation):
+            measurement_qubits.remove(operation.qubits[0])
+            indices_measurement.remove(qubits.index(operation.qubits[0]))
+
+    assert len(indices_measurement) == len(
+        measurement_qubits), 'Indices do not match qubits'
+
+    q_dict = {}
+    for index in indices_measurement:
+        q_dict[measurement_qubits[index]] = index
+
+    # add measurement gate to appropriate qubits
+
+    circuit = measurement_gates(circuit, q_dict)
+
+    # run circuit and store results
+    results = sim.run(circuit, repetitions=n_samples)
+    results_dict = results.measurements
+
+    identity_coeficient = 0
+    expectation_samples = []
+    for rep in range(n_samples):
+        rep_term = rep
+        expectation = 0
+
+        # loop over terms in cost function
+        for term, coef in operator.items():
+            # skip over identity terms
+            if len(term) == 0:
+                identity_coeficient = coef
+                continue
+
+            # here we enforce that hamiltonian is quadratic in pauli.Z
+
+            # no more than 2 terms
+            assert len(term) <= 2, \
+                'quadratic_z flag only accepts quadratic Operators'
+
+            # enforce that operators are pauli.Z
+            # first create a pauli string of Z acting on same qubits
+            z_paulis = {}
+            for qubit in term.keys():
+                z_paulis[qubit] = Pauli.Z
+
+            new_pauli_string = PauliString(qubit_pauli_map=z_paulis)
+
+            # enforce commutativity with new string
+            assert term.commutes_with(new_pauli_string), \
+                'Operator must be diagonal in computational basis' \
+                'and can only contain Pauli.Z is quadratic_z == True'
+
+            # caculates expectation
+
+            q1 = q_dict[list(term.qubits())[0]]
+            if len(term) == 2:
+                q2 = q_dict[list(term.qubits())[1]]
+            else:
+                q2 = q1
+
+            x1 = -2 * int(results_dict[q1][rep_term][0]) + 1
+            x2 = -2 * int(results_dict[q2][rep_term][0]) + 1
+            print(x1,x2)
+
+            expectation += x1 * coef * x2
+
+        expectation_samples.append(expectation)
```

Comment:
Is `np.mean` more numerically stable than computing the total and then dividing by the count? If I specify a billion repetitions, that's quite a large difference in memory.

Reply:
Just computing the total will save memory, I initially had the array so it would be easy to modify and look at a histogram of results. 
I'll push a version where it just computes the sum and divides by the number of samples.